### PR TITLE
Calculate variable from constraint differentiation mode

### LINKED ISSUE #2542: Bug when using solve_strongly_connected_components
## Summary
I have been attempting to apply the `solve_strongly_connected_components` method to the IDAES model libraries and have encountered a couple of bugs the underlying Pyomo code. The first issue that comes up appears to be a missing import in `pyomo/core/base/external.py`, however fixing this unearths a deeper issue.

### Steps to reproduce the issue

The best replication I have is to run the IDAES test suite using the following branch: https://github.com/andrewlee94/idaes-pse/tree/bt_solver

One test that reproduces this error is:

`pytest idaes/models/properties/modular_properties/examples/tests/test_CO2_bmimPF6_PR.py::TestFlashIntegration::test_initialize`

### Error Message

The first bug lies in `external.py`, where it appears that the `value` method needs to be imported from `pyomo.core.expr.numvalue`

```
../../pyomo/pyomo/pyomo/core/base/external.py:214: in evaluate
    args_ = [arg if arg.__class__ in native_types else value(arg)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   args_ = [arg if arg.__class__ in native_types else value(arg) for arg in args]
E   NameError: name 'value' is not defined

../../pyomo/pyomo/pyomo/core/base/external.py:214: NameError
```

However, fixing this results in a second error from deeper in the code:

```
idaes/core/base/unit_model.py:527: in initialize
    return solve_strongly_connected_components(
../../pyomo/pyomo/pyomo/contrib/incidence_analysis/util.py:147: in solve_strongly_connected_components
    results = calculate_variable_from_constraint(
../../pyomo/pyomo/pyomo/util/calc_var_value.py:158: in calculate_variable_from_constraint
    expr_deriv = differentiate(expr, wrt=variable,
../../pyomo/pyomo/pyomo/core/expr/calculus/derivatives.py:116: in differentiate
    res = sympy_diff(expr=expr, wrt=wrt, wrt_list=wrt_list)
../../pyomo/pyomo/pyomo/core/expr/calculus/diff_with_sympy.py:47: in differentiate
    objectMap, sympy_expr = sympyify_expression(expr)
../../pyomo/pyomo/pyomo/core/expr/sympy_tools.py:215: in sympyify_expression
    return object_map, visitor.walk_expression(expr)
../../pyomo/pyomo/pyomo/core/expr/visitor.py:256: in walk_expression
    result = self._process_node(root, RECURSION_LIMIT)
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:455: in _process_node_bx
    data.append(self._process_node(child, recursion_limit))
../../pyomo/pyomo/pyomo/core/expr/visitor.py:466: in _process_node_bx
    return self.exitNode(node, data)
../../pyomo/pyomo/pyomo/core/expr/sympy_tools.py:152: in exitNode
    return node._apply_operation(values)
../../pyomo/pyomo/pyomo/core/expr/numeric_expr.py:650: in _apply_operation
    return self._fcn.evaluate( result )
../../pyomo/pyomo/pyomo/core/base/external.py:216: in evaluate
    return self._evaluate(args_, None, 0)[0]
../../pyomo/pyomo/pyomo/core/base/external.py:360: in _evaluate
    arglist = _ARGLIST(args, fgh, fixed)
../../pyomo/pyomo/pyomo/core/base/external.py:652: in __init__
    self.ra = (c_double*nr)(*_reals)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = 0.120272355044943*x6*(0.6468651916804*x0*x5/x1 + 0.6468651916804*x2*x3/x4)/x7 - 1.0

    def __float__(self):
        # Don't bother testing if it's a number; if it's not this is going
        # to fail, and if it is we still need to check that it evalf'ed to
        # a number.
        result = self.evalf()
        if result.is_Number:
            return float(result)
        if result.is_number and result.as_real_imag()[1]:
            raise TypeError("Cannot convert complex to float")

>       raise TypeError("Cannot convert expression to float")

E       TypeError: Cannot convert expression to float

../../anaconda3/lib/python3.9/site-packages/sympy/core/expr.py:345: TypeError
```

### PR DESCRIPTION
**Pull request by michaelbynum**
Add an option to calculate_variable_from_constraint to use different modes of differentiation in order to support external functions.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/util/tests/test_calc_var_value.py [hunk 1]
```diff
@@ -32,36 +33,57 @@ def test_initialize_value(self):
         m.x.set_value(None)
         calculate_variable_from_constraint(m.x, m.c)
         self.assertEqual(value(m.x), 5)
+        m.x.set_value(None)
+        calculate_variable_from_constraint(m.x, m.c, diff_mode=differentiate.Modes.reverse_numeric)
+        self.assertEqual(value(m.x), 5)
```
Comment (jsiirola) [inline pyomo/util/tests/test_calc_var_value.py:39]: Should we also update these tests for `reverse_symbolic` and explicitly `sympy`?

File: pyomo/util/calc_var_value.py [hunk 2]
```diff
@@ -190,7 +195,13 @@ def calculate_variable_from_constraint(variable, constraint,
                 "expression.\n\tPlease provide a different initial guess "
                 "or enable the linesearch if you have not.")
             raise
-        fpk = value(expr_deriv)
+
+        if diff_mode in {differentiate.Modes.sympy, differentiate.Modes.reverse_symbolic}:
```
Comment (jsiirola) [inline pyomo/util/calc_var_value.py:199]: Note that no test covers the case that `reverse_symbolic` is specified....

Comment (jsiirola) [inline pyomo/util/calc_var_value.py:199]: Note that this test is somewhat expensive, (you have to create the set and then check for membership in it).  I think I would rather do the test once above (line 162) and if it is False, then set `expr_deriv = None`.  Then here we can just do:
```python
if expr_deriv is None:
    fpk = differentiate(expr, wrt=variable, mode=diff_mode)
else:
    fpk = value(expr_deriv)
```

**Review by jsiirola — CHANGES_REQUESTED** [review]
This looks pretty good.  One request to change a potentially expensive test in the inner loop, and some requests for additional tests.

We should also do some profiling and see if numeric is faster than symbolic.  If it is, we should consider changing the default.

**Review by andrewlee94 — COMMENTED** [review]
This looks good to me, however could we also update the `solve_strongly_connected_components` function to support this argument as part of the PR, as it will need a way to set which method to use when solving 1x1 blocks that may contain external functions?

**Comment by Robbybp** [comment]
@andrewlee94 A `calc_var_kwds` argument, which should be a dict of arguments to pass to `calculate_variables_from_constraint`, already exists in `solve_strongly_connected_components`! I added it because at some point I needed to send custom tolerances.

File: pyomo/util/tests/test_calc_var_value.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/util/tests/test_calc_var_value.py]: Done!

File: pyomo/util/calc_var_value.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/util/calc_var_value.py]: Done!

Comment (michaelbynum) [reply pyomo/util/calc_var_value.py]: Done!

**Review by jsiirola — APPROVED** [review]

**Pull request merged**